### PR TITLE
Added unit tests fixed existing errors

### PR DESCRIPTION
**Pull request by hayesgb**
Current PR adds unit tests (bringing total to 22), passes linting and all automated unit tests.

### CODE REVIEW

**Comment by tebeka** [comment]
Thanks @hayesgb! Can you please resolve the conflicts.

File: setup.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ def version():
 
 
 def load_deps(file_name):
-    """Load dependencies from requirements file"""
```
Comment (tebeka) [inline setup.py:26]: Why? These changes makes it hard to focus on the real changes. Also `"""` is [the convention](https://www.python.org/dev/peps/pep-0257/).

File: tests/conftest.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
 
 host = environ.get('V3IO_API')
 access_key = environ.get('V3IO_ACCESS_KEY')
-test_container = 'bigdata'  # TODO: configure
```
Comment (tebeka) [inline tests/conftest.py:27]: Why did you remove the comment?

File: tests/conftest.py [hunk 3]
```diff
@@ -38,7 +38,7 @@ def fs():
     fs._client.close()
 
 
-@pytest.fixture
+@pytest.fixture(scope="session")
```
Comment (tebeka) [inline tests/conftest.py:41]: I don't think session scope is the right one. I want a unique object for each test function.

File: tests/conftest.py [hunk 4]
```diff
@@ -48,26 +48,36 @@ def tmp_obj():
     resp = client.put_object(test_container, path, body=body)
     assert resp.status_code == HTTPStatus.OK, 'create failed'
 
-    yield Obj(f'/{test_container}/{path}', body)
+    body = datetime.now().isoformat().encode('utf-8')
```
Comment (tebeka) [inline tests/conftest.py:51]: - If we add more than one file, need to change the fixture name.
- Use a 'for' loop to create all the files

File: tests/test_file.py [hunk 5]
```diff
@@ -15,23 +15,27 @@
...
-def test_fetch_range(fs: V3ioFS, tmp_obj):
-    v3f = V3ioFile(fs, tmp_obj.path)
-    start, end = 3, len(tmp_obj.data) - 3
+def test_fetch_range(fs: V3ioFS, tmp_file):
+    v3f = V3ioFile(fs, tmp_file.path)
+    start, end = 3, len(tmp_file.data) - 3
     data = v3f._fetch_range(start, end)
-    expected = tmp_obj.data[start:end]
+    expected = tmp_file.data[start:end]
+    print("expected...")
```
Comment (tebeka) [inline tests/test_file.py:23]: These prints are not needed. pytest will show the diff if there's an error.

File: tests/test_fs.py [hunk 6]
```diff
@@ -23,16 +23,77 @@
...
 
 
-def test_ls(fs: V3ioFS, new_file):
-    new_file(fs._client, f'{test_dir}/test-file')  # Make sure dir exists
+def test_ls(fs: V3ioFS, tmp_obj):
+
     path = f'/{test_container}/{test_dir}/'
-    out = fs.ls(path)
-    assert len(out) > 0, 'nothing found'
-    assert all(isinstance(p, dict) for p in out), 'not dict'
```
Comment (tebeka) [inline tests/test_fs.py:31]: Why did you remove this check?

File: tests/test_fs.py [hunk 7]
```diff
@@ -23,16 +23,77 @@
...
+    out0 = fs.ls(path)
+    assert len(out0) > 0, 'nothing found'
+
+    out1 = fs.ls(path, detail=False)
+    assert len(out1) > 0, 'nothing found'
+    assert all(isinstance(p, str) for p in out1), 'not string'
+    out1 = [f for f in out1 if not 'iguazio' in f]
+    assert set(out1) == set(
+        ['/bigdata/v3io-fs-test/test-file',
```
Comment (tebeka) [inline tests/test_fs.py:37]: Constuct the paths from `test_dir`, not hard coded. Even better approach will be to have `tmp_obj` contain the list of files & directrries (probably as a `dict`).

This applies to all cases below, including `test_glob`

File: v3iofs/file.py [hunk 8]
```diff
@@ -15,21 +15,46 @@
...
 from fsspec.spec import AbstractBufferedFile
-from v3io.dataplane import Client
 
 from .path import split_container
 
 
 class V3ioFile(AbstractBufferedFile):
+    """ File-like operation on V3IO Files """
+
+    def __init__(
```
Comment (tebeka) [inline v3iofs/file.py:25]: What's the reason for adding explicit `__init__`? Is it just for direct acces to `client`?

File: v3iofs/fs.py [hunk 9]
```diff
@@ -56,58 +54,70 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
         )
 
-    def ls(self, path, detail=True, **kwargs):
+    @classmethod
+    def _strip_protocol(cls, path):
+        """ Turn path from fully-qualified to file-system-specific
+        May require FS-specific handling, e.g., for relative paths or links.
+        """
+        path = stringify_path(path)
+        protos = (cls.protocol,) if isinstance(cls.protocol, str) else cls.protocol # NOQA
```
Comment (tebeka) [inline v3iofs/fs.py:63]: What does this do?

File: v3iofs/fs.py [hunk 10]
```diff
@@ -56,58 +54,70 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+        for protocol in protos:
```
Comment (tebeka) [inline v3iofs/fs.py:64]: - What are the cases we get the protocol in front?
- I'd like to see some tests for this
- IMO easier done with a regular expression
```
path = re.sub('[a-z0-9]+:(//)?(.*)', r'\2', path)
return '/' + path if path[:0] != '/' else path
```

File: v3iofs/fs.py [hunk 11]
```diff
@@ -56,58 +54,70 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+    def ls(self, path, detail=False, **kwargs):
         """Lists files & directories under path"""
+        path = self._strip_protocol(path)
         full_path = path
         container, path = split_container(path)
         if not container:
             return self._list_containers(detail)
 
-        err = False
-        try:
```
Comment (tebeka) [inline v3iofs/fs.py:67]: Why did you remove the `try/except` block?

File: v3iofs/fs.py [hunk 12]
```diff
@@ -56,58 +54,70 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
-            _resp_dirs(resp, container, detail) +
-            _resp_files(resp, container, detail)
-        )
-
-        if not out:
+        if not pathlist:
             raise FileNotFoundError(f'{full_path!r} not found')
 
-        return out
+        pathlist_ = self._details(pathlist)
```
Comment (tebeka) [inline v3iofs/fs.py:115]:
```
if detail:
    return self._details(pathlist)
return [path['name'] for path in pathlist]
```

File: v3iofs/fs.py [hunk 13]
```diff
@@ -176,14 +187,15 @@ def _open(
         mode='rb',
         block_size=None,
         autocommit=True,
-        cache_options=None,
+        cache_options="readahead",
```
Comment (tebeka) [inline v3iofs/fs.py:190]: What does this do?

File: v3iofs/fs.py [hunk 14]
```diff
@@ -211,39 +214,58 @@ def container_info(container):
...
+    return unslash(f'/{container_name}{prefix}')
+
+
+def prefix_info(container_name, prefix):
+    return info_of(container_name, prefix, 'prefix')
+
 
+def object_path(container_name, object):
+    # object.key already have a leading /
+    return f'/{container_name}{object.key}'
```
Comment (tebeka) [inline v3iofs/fs.py:233]: I think this is very confusing, in previous commit I made sure `object` does *not* have a leading `/'.
If this is an issue, I prefer to have `f'{container_name}/{object.key}' and then normalize it.

File: v3iofs/fs.py [hunk 15]
```diff
@@ -211,39 +214,58 @@ def container_info(container):
...
-    return f'/{container}/{path}'
+
+def object_info(container_name, object):
+    return info_of(container_name, object, 'key')
 
 
 def info_of(container_name, obj, name_key):
     info = {
-        'name': obj_path(container_name, obj, name_key),
-        'type': 'directory' if hasattr(obj, 'size') else 'file',
```
Comment (tebeka) [inline v3iofs/fs.py:233]: Where did `type` go?

**Review by tebeka — CHANGES_REQUESTED** [review]
Why were some test files renamed to have a leading `_`?
I don't understand what is fixed by this PR. Please try to have small PRs with well defined functionality.

File: setup.py (hunk 1 above)
Comment (hayesgb) [reply setup.py:26]: Updated.

File: tests/conftest.py (hunk 2 above)
Comment (hayesgb) [reply tests/conftest.py:27]: Added comment back

File: tests/conftest.py (hunk 3 above)
Comment (hayesgb) [reply tests/conftest.py:41]: Updated scope to function.

File: tests/test_file.py (hunk 5 above)
Comment (hayesgb) [reply tests/test_file.py:23]: Removed per request.

File: v3iofs/file.py (hunk 8 above)
Comment (hayesgb) [reply v3iofs/file.py:25]: Yes.  It also explicitly allows others to understand the declared mode, block_size, autocommit behavior, cache_type, and cache_options here, rather than having to go look at the AbstractBuffered file.

File: tests/test_fs.py (hunk 6 above)
Comment (hayesgb) [reply tests/test_fs.py:31]: The original test included a new_file object created in conftest, and new_file only included a single file one layer deep.  This was removed, and tmp_obj used to be consistent with other tests.  The additional file cases below were added to validate ability to list nested files, and to verify that file details are properly identified.

File: v3iofs/fs.py (hunk 9 above)
Comment (hayesgb) [reply v3iofs/fs.py:63]: Checks to verify that a path is passed as a string.  If its passed as a filesystem object, attempts to convert it to a string.  Also identifies the protocol.  In this case, v3io

File: v3iofs/fs.py (hunk 10 above)
Comment (hayesgb) [reply v3iofs/fs.py:64]: Works if the path is passed as a string.  If it's passed as a pathlib object, will this work?  At any rate, the code below follows the same convention as what already exists in the fsspec library.

File: v3iofs/fs.py (hunk 11 above)
Comment (hayesgb) [reply v3iofs/fs.py:67]: As written, the code demands the return of a HTTP.OK status.  I found it more straightforward to NOT raise an exception based on the HTTP response, but rather to inspect the response for the presence of both directories and files.

File: v3iofs/fs.py (hunk 12 above)
Comment (hayesgb) [reply v3iofs/fs.py:115]: Updated.  Linter requires a linespace.

File: v3iofs/fs.py (hunk 13 above)
Comment (hayesgb) [reply v3iofs/fs.py:190]: fsspec provides several caching methods.  The default is readahead.  See:  https://github.com/intake/filesystem_spec/blob/master/fsspec/caching.py

File: v3iofs/fs.py (hunk 15 above)
Comment (hayesgb) [reply v3iofs/fs.py:233]: It is set in self._details_of, which handles defining files and directories

**Comment by hayesgb** [comment]
> Why were some test files renamed to have a leading `_`?
> I don't understand what is fixed by this PR. Please try to have small PRs with well defined functionality.

@tebeka -- This PR is intended to allow expected functionality for reading/writing to parquet files with Dask.  To validate this functionality, several tests were added, such that, if passing, should validate this requirement.  The leading "_" were added, because executing some of the tests in my local environment (rather than the v3io filesystem) does not proceed as expected.  This extra layer of complexity leads to having to do some of my development within the Iguazio platform, where I do not have access to a full-features IDE, but rather works gets done in the Jupyter editor.

**Comment by tebeka** [comment]
Thanks @hayesgb - I'll incorporate this PR to master soon.

**Comment by tebeka** [comment]
Closing this one - incorperated changes in `master`.
If there are more demends please open an issue - thanks!
